tools/windows-node-installer/test/e2e/azure_test.go

```diff
@@ -258,27 +280,25 @@ ansible_winrm_server_cert_validation=ignore`, ip, password, winRMPort))
 
 // testAnsiblePing checks if ansible is able to ping on opened winRmHttps port
 func testAnsiblePing(t *testing.T) {
-	// this regex looks for the IP address pattern from vmCredentialPath.
-	// the sample vmCredentialPath looks like xfreerdp /u:xxxx /v:xx.xx.xx.xx /h:1080 /w:1920 /p:'password1234'
-	ipAddressPattern := regexp.MustCompile(`\d+.\d+.\d+.\d+`)
-	// the passwordPattern extracts the characters present after the '/p:', but we are extracting
-	// 13 characters even though the windows password length is of size 12 because we got a single quote
-	// character included in the password.
-	passwordPattern := regexp.MustCompile(`/p:.{13}`)
 	for _, vmName := range instanceIDs {
-		vmCredentialPath := filepath.Join(dir, "/", vmName)
-		rdpCmd, err := ioutil.ReadFile(vmCredentialPath)
+		ipAddress, password, err := getIpPass(vmName)
 		require.NoErrorf(t, err, "failed to read file %s", vmName)
-		ipAddress := ipAddressPattern.FindString(string(rdpCmd))
 		assert.NotEmpty(t, ipAddress, "the IP address can't be empty")
-		password := passwordPattern.FindString(string(rdpCmd))[3:]
 		assert.NotEmpty(t, password, "the password can't be empty")
-		// we are trimming out the unnecessary single quotes.
-		password = strings.Trim(password, `'`)
 		hostFileName, err := createHostFile(ipAddress, password)
 		require.NoError(t, err, "failed to create a temp file")
 		pingCmd := exec.Command("ansible", "win", "-i", hostFileName, "-m", "win_ping")
 		out, err := pingCmd.CombinedOutput()
 		assert.NoErrorf(t, err, "ansible ping check failed with error: %s", string(out))
 	}
 }
+
+// testAzureFirewallRule asserts if the created instance has firewall rule that opens kubelet port.
+func testAzureInstancesFirewallRule(t *testing.T) {
+	for _, vmName := range instanceIDs {
+		ipAddress, password, err := getIpPass(vmName)
+		require.NoErrorf(t, err, "failed to obtain credentials for %s", vmName)
+		credentials = types.NewCredentials(vmName, ipAddress, password, azureUser)
```

Comment:
You're setting a global variable here in a loop. Why not just pass in the credentials to the `testInstanceFirewallRule` function?

Reply:
I second @sebsoto on this. We are in this weird spot where we assume there is only one VM but loop through these instances in many places. So we either stop looping or have a better way to pass the credentials around. This is going to nasty to unravel if we ever decide to run VMs in our WNI e2e tests.